File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/task/TaskManager.kt
package com.coinbase.android.nativesdk.task

import com.coinbase.android.nativesdk.CoinbaseWalletSDKError
import com.coinbase.android.nativesdk.message.request.UnencryptedRequestMessage
import com.coinbase.android.nativesdk.message.response.ResponseContent
import com.coinbase.android.nativesdk.message.response.ResponseHandler
import com.coinbase.android.nativesdk.message.response.ResponseResult
import com.coinbase.android.nativesdk.message.response.UnencryptedResponseMessage

internal class TaskManager {
    private val tasks = HashMap<String, Task>()

    fun registerResponseHandler(message: UnencryptedRequestMessage, handler: ResponseHandler) {
        tasks[message.uuid] = Task(
            request = message,
            handler = handler,
            timestamp = message.timestamp
        )
    }

    fun handleResponse(message: UnencryptedResponseMessage): Boolean {
        val requestId: String
        val result: ResponseResult = when (val response = message.content.sealed) {
            is ResponseContent.Response -> {
                requestId = response.requestId
                Result.success(response.values)
            }
            is ResponseContent.Failure -> {
                requestId = response.requestId
                Result.failure(CoinbaseWalletSDKError.WalletReturnedError(response.description))
            }
        }

        val task = tasks[requestId] ?: return false

        task.handler(result)
        tasks.remove(requestId)
        return true
    }

    fun reset(){
        tasks.clear()
    }
}

File: ios/CoinbaseWalletSDK/Task/TaskManager.swift
//
//  TaskManager.swift
//  MobileWalletProtocol
//
//  Created by Jungho Bang on 6/9/22.
//

import Foundation

@available(iOS 13.0, *)
class TaskManager {
    private static var tasks = [UUID: Task]()
    
    static func registerResponseHandler(
        for request: RequestMessage,
        host: URL,
        _ handler: @escaping ResponseHandler
    ) {
        let uuid = request.uuid
        tasks[uuid] = Task(
            request: request,
            host: host,
            handler: handler,
            timestamp: Date()
        )
    }
    
    @discardableResult static func runResponseHandler(with response: ResponseMessage) -> Bool {
        let requestId = response.content.requestId
 
        guard let task = tasks[requestId] else {
            return false
        }
        
        task.handler(response.result)
        tasks.removeValue(forKey: requestId)
        return true
    }
    
    static func findTask(for requestId: UUID) -> Task? {
        return tasks[requestId]
    }
    
    static func reset(host: URL) {
        tasks = tasks.filter { $0.value.host != host }
    }
    
}
